datafusion/physical-expr/src/simplifier/not.rs

```diff
@@ -0,0 +1,379 @@
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_demorgans_law_or() -> Result<()> {
+        let schema = test_schema();
+
+        // NOT(a OR b) -> NOT a AND NOT b
+        let or_expr = Arc::new(BinaryExpr::new(
+            col("a", &schema)?,
+            Operator::Or,
+            col("b", &schema)?,
+        ));
+        let not_or: Arc<dyn PhysicalExpr> = Arc::new(NotExpr::new(or_expr));
+
+        let result = simplify_not_expr_recursive(&not_or, &schema)?;
+        assert!(result.transformed);
+
+        if let Some(binary) = result.data.as_any().downcast_ref::<BinaryExpr>() {
+            assert_eq!(binary.op(), &Operator::And);
+            // Left and right should both be NOT expressions
+            assert!(binary.left().as_any().downcast_ref::<NotExpr>().is_some());
+            assert!(binary.right().as_any().downcast_ref::<NotExpr>().is_some());
+        } else {
+            panic!("Expected binary expression result");
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_demorgans_with_comparison_simplification() -> Result<()> {
+        let schema = test_schema();
+
+        // NOT(b = 1 AND b = 2) -> b != 1 OR b != 2
+        // This tests the combination of De Morgan's law and operator negation
+        let eq1 = Arc::new(BinaryExpr::new(
+            col("b", &schema)?,
+            Operator::Eq,
+            lit(ScalarValue::Int32(Some(1))),
+        ));
+        let eq2 = Arc::new(BinaryExpr::new(
+            col("b", &schema)?,
+            Operator::Eq,
+            lit(ScalarValue::Int32(Some(2))),
+        ));
+        let and_expr = Arc::new(BinaryExpr::new(eq1, Operator::And, eq2));
+        let not_and: Arc<dyn PhysicalExpr> = Arc::new(NotExpr::new(and_expr));
+
+        let result = simplify_not_expr_recursive(&not_and, &schema)?;
+        assert!(result.transformed, "Expression should be transformed");
+
+        // Verify the result is an OR expression
+        if let Some(or_binary) = result.data.as_any().downcast_ref::<BinaryExpr>() {
+            assert_eq!(or_binary.op(), &Operator::Or, "Top level should be OR");
+
+            // Verify left side is b != 1
+            if let Some(left_binary) =
+                or_binary.left().as_any().downcast_ref::<BinaryExpr>()
+            {
+                assert_eq!(left_binary.op(), &Operator::NotEq, "Left should be NotEq");
+            } else {
+                panic!("Expected left to be a binary expression with !=");
+            }
+
+            // Verify right side is b != 2
+            if let Some(right_binary) =
+                or_binary.right().as_any().downcast_ref::<BinaryExpr>()
+            {
+                assert_eq!(right_binary.op(), &Operator::NotEq, "Right should be NotEq");
+            } else {
+                panic!("Expected right to be a binary expression with !=");
+            }
+        } else {
+            panic!("Expected binary OR expression result");
+        }
+
+        Ok(())
```

Comment:
Please add a test for `NOT(NOT(a) AND NOT(b))`. 
It should be simplified to `a OR b`

Reply:
done https://github.com/apache/datafusion/pull/18970/commits/a0116cec5edbd576440610fe4d137f784235e4e0